infobind/cache.go

```diff
@@ -0,0 +1,215 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package infobind
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/parser"
+	"github.com/pingcap/parser/ast"
+	"github.com/pingcap/parser/terror"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/sqlexec"
+	log "github.com/sirupsen/logrus"
+)
+
+const defaultBindCacheSize = 5
+
+// bindData store the basic bind info and bindSql astNode.
+type bindData struct {
+	bindRecord
+	ast ast.StmtNode
+}
+
+// bindCache hold a bindDataMap, key:origin sql hash value: bindData slice.
+type bindCache struct {
+	Cache map[string][]*bindData
+}
+
+// Handle hold a atomic bindCache.
+type Handle struct {
```

Comment:
1. Can we define Handle as 
``` go
type Handler struct{
  *sync.Mutex
  cache bindCache
}
```
2. Add a Set func for Handle
3. Can we define bindCache as `type bindCache = map[string][]*bindData`
4. Can we rename Handle as Handler?

Reply:
why we replace CAS operator with lock? What's advantage of mutex?